apisix/plugins/authz-casbin.lua

```diff
@@ -0,0 +1,347 @@
+
+        if casbin_enforcer:RemovePolicy(subject, object, action) then
+            local ok, _ = save_policy()
+            if ok == 400 then
+                return 400, {message = "Failed to save policy, see logs."}
+            end
+            return 200, {message = "Successfully removed policy."}
+        else
+            return 400, {message = "Invalid policy request."}
+        end
+    elseif type == "g" then
+        local user = headers["user"]
+        local role = headers["role"]
+
+        if not user or not role then
+            return 400, {message = "Invalid policy request."}
+        end
+
+        if casbin_enforcer:RemoveGroupingPolicy(user, role) then
+            local ok, _ = save_policy()
+            if ok == 400 then
+                return 400, {message = "Failed to save policy, see logs."}
+            end
+            return 200, {message = "Successfully removed grouping policy."}
+        else
+            return 400, {message = "Invalid policy request."}
+        end
+    else
+        return 400, {message = "Invalid policy type."}
+    end
+end
+
+
+local function has_policy()
+    if not casbin_enforcer then
+        return 400, {message = "Enforcer not created yet."}
+    end
+
+    local headers = get_headers()
+    local type = headers["type"]
+
+    if type == "p" then
+        local subject = headers["subject"]
+        local object = headers["object"]
+        local action = headers["action"]
+
+        if not subject or not object or not action then
+            return 400, {message = "Invalid policy request."}
+        end
+
+        if casbin_enforcer:HasPolicy(subject, object, action) then
+            return 200, {data = "true"}
+        else
+            return 200, {data = "false"}
+        end
+    elseif type == "g" then
+        local user = headers["user"]
+        local role = headers["role"]
+
+        if not user or not role then
+            return 400, {message = "Invalid policy request."}
+        end
+
+        if casbin_enforcer:HasGroupingPolicy(user, role) then
+            return 200, {data = "true"}
+        else
+            return 200, {data = "false"}
+        end
+    else
+        return 400, {message = "Invalid policy type."}
+    end
+end
+
+
+local function get_policy()
+    if not casbin_enforcer then
+        return 400, {message = "Enforcer not created yet."}
+    end
+
+    local headers = get_headers()
+    local type = headers["type"]
+
+    if type == "p" then
+        local policy = casbin_enforcer:GetPolicy()
+        if policy then
+            return 200, {data = policy}
+        else
+            return 400
+        end
+    elseif type == "g" then
+        local groupingPolicy = casbin_enforcer:GetGroupingPolicy()
+        if groupingPolicy then
+            return 200, {data = groupingPolicy}
+        else
+            return 400
+        end
+    else
+        return 400, {message = "Invalid policy type."}
+    end
+end
+
+
+function _M.api()
+    return {
```

Comment:
Since we already use metadata to store the configuration, there is no need to provide APIs on the DP side. Everything should be done with the plugin metadata's API on the CP side.

Reply:
@spacewander 
Thanks for your review, so we will remove all the API functions and if the model/policy text in plugin metadata is updated (from the CP side), we will update the `CasbinEnforcer` through that model/policy too. Is this correct?

Reply:
Yes, you can use `metadata.modifiedIndex` to ensure casbin_enforcer is created with the latest model/policy, like this one:
https://github.com/apache/apisix/blob/fd8f875429bd43462430bd662788940f6a579ff9/apisix/http/router/radixtree_uri.lua#L31-L36

Reply:
And I think you can submit a minimum viable pull request so that we can check & merge it early.

Reply:
Sure I will do that.

Reply:
I added a condition that if the schema changes, the `casbin_enforcer` will be recreated.